File: AccuPay/Reports/Reports/DataModel/LeaveLedgerReportModel.vb
Option Strict On

Imports AccuPay.Core
Imports AccuPay.Core.Enums
Imports AccuPay.Core.Helpers

''' <summary>
''' Anemic implementation of ILeaveLedgerReportModel just for Crystal Report data source
''' </summary>
Public Class LeaveLedgerReportModel
    Implements ILeaveLedgerReportModel

    Public ReadOnly Property EmployeeNumber As String Implements ILeaveLedgerReportModel.EmployeeNumber

    Public ReadOnly Property FullName As String Implements ILeaveLedgerReportModel.FullName

    Public ReadOnly Property LeaveType As LeaveType Implements ILeaveLedgerReportModel.LeaveType

    Sub New(employeeNumber As String,
            fullName As String,
            leaveType As LeaveType,
            beginningBalance As Decimal,
            availedLeave As Decimal)

        Me.EmployeeNumber = employeeNumber
        Me.FullName = fullName
        Me.LeaveType = leaveType
        Me.BeginningBalance = beginningBalance
        Me.AvailedLeave = availedLeave
    End Sub

    Public ReadOnly Property LeaveTypeDescription As String Implements ILeaveLedgerReportModel.LeaveTypeDescription
        Get
            Select Case LeaveType
                Case LeaveType.Sick
                    Return "SL"
                Case LeaveType.Vacation
                    Return "VL"
                Case Else
                    Return ""
            End Select
        End Get

    End Property

    Public ReadOnly Property BeginningBalance() As Decimal Implements ILeaveLedgerReportModel.BeginningBalance

    Public ReadOnly Property AvailedLeave() As Decimal Implements ILeaveLedgerReportModel.AvailedLeave

    Public ReadOnly Property BeginningBalanceInDays As Decimal Implements ILeaveLedgerReportModel.BeginningBalanceInDays
        Get
            Return BeginningBalance / Helpers.PayrollTools.WorkHoursPerDay
        End Get
    End Property

    Public ReadOnly Property AvailedLeaveInDays As Decimal Implements ILeaveLedgerReportModel.AvailedLeaveInDays
        Get
            Return AvailedLeave / PayrollTools.WorkHoursPerDay
        End Get
    End Property

    Public ReadOnly Property EndingBalance As Decimal Implements ILeaveLedgerReportModel.EndingBalance
        Get
            Return BeginningBalance - AvailedLeave
        End Get
    End Property

    Public ReadOnly Property EndingBalanceInDays As Decimal Implements ILeaveLedgerReportModel.EndingBalanceInDays
        Get
            Return EndingBalance / 8
        End Get
    End Property

End Class

File: AccuPay.Core/ReportModels/LeaveLedgerReportModel.cs
using AccuPay.Core.Enums;
using AccuPay.Core.Helpers;
using AccuPay.Utilities;

namespace AccuPay.Core.ReportModels
{
    public class LeaveLedgerReportModel : ILeaveLedgerReportModel
    {
        public string EmployeeNumber { get; set; }

        public string FullName { get; set; }

        public LeaveType LeaveType { get; set; }

        public LeaveLedgerReportModel(string employeeNumber,
                                        string fullName,
                                        LeaveType leaveType,
                                        decimal beginningBalance,
                                        decimal availedLeave)
        {
            this.EmployeeNumber = employeeNumber;
            this.FullName = fullName;
            this.LeaveType = leaveType;
            this.BeginningBalance = beginningBalance;
            this.AvailedLeave = availedLeave;
        }

        public string LeaveTypeDescription
        {
            get
            {
                switch (LeaveType)
                {
                    case LeaveType.Sick:
                        return "SL";

                    case LeaveType.Vacation:
                        return "VL";

                    default:
                        return "";
                }
            }
        }

        private decimal _beginningBalance;

        public decimal BeginningBalance
        {
            get => _beginningBalance;
            set => _beginningBalance = AccuMath.CommercialRound(value);
        }

        private decimal _availedLeave;

        public decimal AvailedLeave
        {
            get => _availedLeave;
            set => _availedLeave = AccuMath.CommercialRound(value);
        }

        public decimal BeginningBalanceInDays => BeginningBalance / PayrollTools.WorkHoursPerDay;

        public decimal AvailedLeaveInDays => AvailedLeave / PayrollTools.WorkHoursPerDay;

        public decimal EndingBalance => BeginningBalance - AvailedLeave;

        public decimal EndingBalanceInDays => EndingBalance / 8;
    }
}